From 16c45527aa3566c15ccdf6d82267ec41a21fa895 Mon Sep 17 00:00:00 2001
From: Brendan Cunningham <bcunningham@cornelisnetworks.com>
Date: Wed, 22 Apr 2026 17:17:43 -0400
Subject: [PATCH] Restore pnet/opa component

pnet/opa was removed because of a report that OPA-PSM2 was being
selected with non-OPA devices. Reportedly,
(vendor=0x8086,class=0x0208) had been recycled for a non-OPA100 device.

However this report seems to have been in error; the user did in-fact
have an OPA100 device in their system.

So restore pnet/opa by reverting commit d52ebb7968e6
("Eliminate stale pnet/opa component").

Signed-off-by: Brendan Cunningham <bcunningham@cornelisnetworks.com>
---
 src/mca/pnet/opa/Makefile.am          |  56 ++++
 src/mca/pnet/opa/configure.m4         |  27 ++
 src/mca/pnet/opa/pnet_opa.c           | 400 ++++++++++++++++++++++++++
 src/mca/pnet/opa/pnet_opa.h           |  41 +++
 src/mca/pnet/opa/pnet_opa_component.c | 115 ++++++++
 5 files changed, 639 insertions(+)
 create mode 100644 src/mca/pnet/opa/Makefile.am
 create mode 100644 src/mca/pnet/opa/configure.m4
 create mode 100644 src/mca/pnet/opa/pnet_opa.c
 create mode 100644 src/mca/pnet/opa/pnet_opa.h
 create mode 100644 src/mca/pnet/opa/pnet_opa_component.c

diff --git a/src/mca/pnet/opa/Makefile.am b/src/mca/pnet/opa/Makefile.am
new file mode 100644
index 0000000000..e000896eed
--- /dev/null
+++ b/src/mca/pnet/opa/Makefile.am
@@ -0,0 +1,56 @@
+# -*- makefile -*-
+#
+# Copyright (c) 2004-2005 The Trustees of Indiana University and Indiana
+#                         University Research and Technology
+#                         Corporation.  All rights reserved.
+# Copyright (c) 2004-2005 The University of Tennessee and The University
+#                         of Tennessee Research Foundation.  All rights
+#                         reserved.
+# Copyright (c) 2004-2005 High Performance Computing Center Stuttgart,
+#                         University of Stuttgart.  All rights reserved.
+# Copyright (c) 2004-2005 The Regents of the University of California.
+#                         All rights reserved.
+# Copyright (c) 2012      Los Alamos National Security, Inc.  All rights reserved.
+# Copyright (c) 2013-2019 Intel, Inc.  All rights reserved.
+# Copyright (c) 2017      Research Organization for Information Science
+#                         and Technology (RIST). All rights reserved.
+# Copyright (c) 2021-2022 Nanook Consulting.  All rights reserved.
+# $COPYRIGHT$
+#
+# Additional copyrights may follow
+#
+# $HEADER$
+#
+
+headers = pnet_opa.h
+sources = \
+        pnet_opa_component.c \
+        pnet_opa.c
+
+# Make the output library in this directory, and name it either
+# mca_<type>_<name>.la (for DSO builds) or libmca_<type>_<name>.la
+# (for static builds).
+
+if MCA_BUILD_pmix_pnet_opa_DSO
+lib =
+lib_sources =
+component = pmix_mca_pnet_opa.la
+component_sources = $(headers) $(sources)
+else
+lib = libpmix_mca_pnet_opa.la
+lib_sources = $(headers) $(sources)
+component =
+component_sources =
+endif
+
+mcacomponentdir = $(pmixlibdir)
+mcacomponent_LTLIBRARIES = $(component)
+pmix_mca_pnet_opa_la_SOURCES = $(component_sources)
+pmix_mca_pnet_opa_la_LDFLAGS = -module -avoid-version
+if NEED_LIBPMIX
+pmix_mca_pnet_opa_la_LIBADD = $(top_builddir)/src/libpmix.la
+endif
+
+noinst_LTLIBRARIES = $(lib)
+libpmix_mca_pnet_opa_la_SOURCES = $(lib_sources)
+libpmix_mca_pnet_opa_la_LDFLAGS = -module -avoid-version
diff --git a/src/mca/pnet/opa/configure.m4 b/src/mca/pnet/opa/configure.m4
new file mode 100644
index 0000000000..f28f445f04
--- /dev/null
+++ b/src/mca/pnet/opa/configure.m4
@@ -0,0 +1,27 @@
+# -*- shell-script -*-
+#
+# Copyright (c) 2021      Nanook Consulting.  All rights reserved.
+# Copyright (c) 2022      Amazon.com, Inc. or its affiliates.
+#                         All Rights reserved.
+# $COPYRIGHT$
+#
+# Additional copyrights may follow
+#
+# $HEADER$
+#
+
+# MCA_pmix_pnet_opa_CONFIG([action-if-can-compile],
+#                          [action-if-cant-compile])
+# ------------------------------------------------
+AC_DEFUN([MCA_pmix_pnet_opa_CONFIG], [
+    AC_CONFIG_FILES([src/mca/pnet/opa/Makefile])
+
+    AS_IF([test "yes" = "yes"],
+          [$1
+           pmix_pnet_opa=yes],
+          [$2
+           pmix_pnet_opa=no])
+
+    PMIX_SUMMARY_ADD([Transports], [OmniPath], [], [$pmix_pnet_opa])
+
+])dnl
diff --git a/src/mca/pnet/opa/pnet_opa.c b/src/mca/pnet/opa/pnet_opa.c
new file mode 100644
index 0000000000..a27a767e13
--- /dev/null
+++ b/src/mca/pnet/opa/pnet_opa.c
@@ -0,0 +1,400 @@
+/*
+ * Copyright (c) 2015-2020 Intel, Inc.  All rights reserved.
+ * Copyright (c) 2016      IBM Corporation.  All rights reserved.
+ *
+ * Copyright (c) 2021-2025 Nanook Consulting  All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "src/include/pmix_config.h"
+
+#include <string.h>
+#ifdef HAVE_UNISTD_H
+#    include <unistd.h>
+#endif
+#ifdef HAVE_SYS_TYPES_H
+#    include <sys/types.h>
+#endif
+#ifdef HAVE_SYS_STAT_H
+#    include <sys/stat.h>
+#endif
+#ifdef HAVE_FCNTL_H
+#    include <fcntl.h>
+#endif
+#include <time.h>
+
+#include "pmix_common.h"
+
+#include "src/class/pmix_list.h"
+#include "src/hwloc/pmix_hwloc.h"
+#include "src/include/pmix_globals.h"
+#include "src/include/pmix_socket_errno.h"
+#include "src/mca/base/pmix_mca_base_var.h"
+#include "src/mca/pcompress/pcompress.h"
+#include "src/mca/preg/preg.h"
+#include "src/util/pmix_alfg.h"
+#include "src/util/pmix_argv.h"
+#include "src/util/pmix_error.h"
+#include "src/util/pmix_name_fns.h"
+#include "src/util/pmix_output.h"
+#include "src/util/pmix_printf.h"
+#include "src/util/pmix_environ.h"
+
+#include "pnet_opa.h"
+#include "src/mca/pnet/base/base.h"
+#include "src/mca/pnet/pnet.h"
+
+static pmix_status_t allocate(pmix_namespace_t *nptr,
+                              pmix_info_t info[], size_t ninfo,
+                              pmix_list_t *ilist);
+static pmix_status_t setup_local_network(pmix_nspace_env_cache_t *nptr,
+                                         pmix_info_t info[], size_t ninfo);
+static pmix_status_t collect_inventory(pmix_info_t directives[], size_t ndirs,
+                                       pmix_list_t *inventory);
+static pmix_status_t deliver_inventory(pmix_info_t info[], size_t ninfo,
+                                       pmix_info_t directives[], size_t ndirs);
+pmix_pnet_module_t pmix_opa_module = {
+    .name = "opa",
+    .allocate = allocate,
+    .setup_local_network = setup_local_network,
+    .collect_inventory = collect_inventory,
+    .deliver_inventory = deliver_inventory
+};
+
+/* some network transports require a little bit of information to
+ * "pre-condition" them - i.e., to setup their individual transport
+ * connections so they can generate their endpoint addresses. This
+ * function provides a means for doing so. The resulting info is placed
+ * into the app_context's env array so it will automatically be pushed
+ * into the environment of every MPI process when launched.
+ */
+
+static inline void transports_use_rand(uint64_t *unique_key)
+{
+    pmix_rng_buff_t rng;
+    pmix_srand(&rng, (unsigned int) time(NULL));
+    unique_key[0] = pmix_rand(&rng);
+    unique_key[1] = pmix_rand(&rng);
+}
+
+static char *transports_print(uint64_t *unique_key)
+{
+    unsigned int *int_ptr;
+    size_t i, j, string_key_len, written_len;
+    char *string_key = NULL, *format = NULL;
+
+    /* string is two 64 bit numbers printed in hex with a dash between
+     * and zero padding.
+     */
+    string_key_len = (sizeof(uint64_t) * 2) * 2 + strlen("-") + 1;
+    string_key = (char *) malloc(string_key_len);
+    if (NULL == string_key) {
+        return NULL;
+    }
+
+    string_key[0] = '\0';
+    written_len = 0;
+
+    /* get a format string based on the length of an unsigned int.  We
+     * want to have zero padding for sizeof(unsigned int) * 2
+     * characters -- when printing as a hex number, each byte is
+     * represented by 2 hex characters.  Format will contain something
+     * that looks like %08lx, where the number 8 might be a different
+     * number if the system has a different sized long (8 would be for
+     * sizeof(int) == 4)).
+     */
+    if (0 > asprintf(&format, "%%0%dx", (int) (sizeof(unsigned int)) * 2)) {
+        free(string_key);
+        return NULL;
+    }
+
+    /* print the first number */
+    int_ptr = (unsigned int *) &unique_key[0];
+    for (i = 0; i < sizeof(uint64_t) / sizeof(unsigned int); ++i) {
+        if (0 == int_ptr[i]) {
+            /* inject some energy */
+            for (j = 0; j < sizeof(unsigned int); j++) {
+                int_ptr[i] |= j << j;
+            }
+        }
+        pmix_snprintf(string_key + written_len, string_key_len - written_len, format, int_ptr[i]);
+        written_len = strlen(string_key);
+    }
+
+    /* print the middle dash */
+    pmix_snprintf(string_key + written_len, string_key_len - written_len, "-");
+    written_len = strlen(string_key);
+
+    /* print the second number */
+    int_ptr = (unsigned int *) &unique_key[1];
+    for (i = 0; i < sizeof(uint64_t) / sizeof(unsigned int); ++i) {
+        if (0 == int_ptr[i]) {
+            /* inject some energy */
+            for (j = 0; j < sizeof(unsigned int); j++) {
+                int_ptr[i] |= j << j;
+            }
+        }
+        pmix_snprintf(string_key + written_len, string_key_len - written_len, format, int_ptr[i]);
+        written_len = strlen(string_key);
+    }
+    free(format);
+
+    return string_key;
+}
+
+/* NOTE: if there is any binary data to be transferred, then
+ * this function MUST pack it for transport as the host will
+ * not know how to do so */
+static pmix_status_t allocate(pmix_namespace_t *nptr, pmix_info_t info[], size_t ninfo,
+                              pmix_list_t *ilist)
+{
+    uint64_t unique_key[2];
+    char *string_key;
+    int fd_rand;
+    size_t bytes_read, n, m, p;
+    pmix_buffer_t mydata; // Buffer used to store information to be transmitted (scratch storage)
+    pmix_kval_t *kv;
+    pmix_envar_t envar;
+    pmix_byte_object_t bo;
+    bool envars = false, seckeys = false;
+    pmix_status_t rc;
+    pmix_info_t *iptr;
+    pmix_list_t cache;
+
+    pmix_output_verbose(2, pmix_pnet_base_framework.framework_output,
+                        "pnet:opa:allocate for nspace %s", nptr->nspace);
+
+    if (NULL == info) {
+        return PMIX_ERR_TAKE_NEXT_OPTION;
+    }
+
+    PMIX_ENVAR_CONSTRUCT(&envar);
+
+    for (n = 0; n < ninfo; n++) {
+        if (PMIX_CHECK_KEY(&info[n], PMIX_SETUP_APP_ENVARS)) {
+            envars = PMIX_INFO_TRUE(&info[n]);
+        } else if (PMIX_CHECK_KEY(&info[n], PMIX_SETUP_APP_ALL)) {
+            envars = PMIX_INFO_TRUE(&info[n]);
+            seckeys = PMIX_INFO_TRUE(&info[n]);
+        } else if (PMIX_CHECK_KEY(&info[n], PMIX_SETUP_APP_NONENVARS)) {
+            seckeys = PMIX_INFO_TRUE(&info[n]);
+        } else if (PMIX_CHECK_KEY(&info[n], PMIX_ALLOC_NETWORK)) {
+            iptr = (pmix_info_t *) info[n].value.data.darray->array;
+            m = info[n].value.data.darray->size;
+            for (p = 0; p < m; p++) {
+                if (PMIX_CHECK_KEY(&iptr[p], PMIX_ALLOC_NETWORK_SEC_KEY)) {
+                    seckeys = PMIX_INFO_TRUE(&iptr[p]);
+                } else if (PMIX_CHECK_KEY(&iptr[p], PMIX_ALLOC_NETWORK_ID)) {
+                    /* need to track the request by this ID */
+                } else if (PMIX_CHECK_KEY(&iptr[p], PMIX_SETUP_APP_ENVARS)) {
+                    envars = PMIX_INFO_TRUE(&iptr[p]);
+                } else if (PMIX_CHECK_KEY(&iptr[p], PMIX_SETUP_APP_ALL)) {
+                    envars = PMIX_INFO_TRUE(&iptr[p]);
+                    seckeys = PMIX_INFO_TRUE(&iptr[p]);
+                } else if (PMIX_CHECK_KEY(&iptr[p], PMIX_SETUP_APP_NONENVARS)) {
+                    seckeys = PMIX_INFO_TRUE(&iptr[p]);
+                }
+            }
+        }
+    }
+    /* setup a buffer - we will pack the info into it for transmission to
+     * the backend compute node daemons */
+    PMIX_CONSTRUCT(&mydata, pmix_buffer_t);
+
+    if (seckeys) {
+        pmix_output_verbose(2, pmix_pnet_base_framework.framework_output,
+                            "pnet: opa providing seckeys");
+        /* put the number here - or else create an appropriate string. this just needs to
+         * eventually be a string variable
+         */
+        if (-1 == (fd_rand = open("/dev/urandom", O_RDONLY))) {
+            transports_use_rand(unique_key);
+        } else {
+            bytes_read = read(fd_rand, (char *) unique_key, 16);
+            if (bytes_read != 16) {
+                transports_use_rand(unique_key);
+            }
+            close(fd_rand);
+        }
+
+        if (NULL == (string_key = transports_print(unique_key))) {
+            PMIX_ERROR_LOG(PMIX_ERR_OUT_OF_RESOURCE);
+            PMIX_DESTRUCT(&mydata);
+            return PMIX_ERR_OUT_OF_RESOURCE;
+        }
+
+        PMIX_ENVAR_LOAD(&envar, "OMPI_MCA_orte_precondition_transports", string_key, ':');
+        PMIX_BFROPS_PACK(rc, pmix_globals.mypeer, &mydata, &envar, 1, PMIX_ENVAR);
+        free(string_key);
+        PMIX_ENVAR_DESTRUCT(&envar);
+    }
+
+    if (envars) {
+        pmix_output_verbose(2, pmix_pnet_base_framework.framework_output,
+                            "pnet: opa harvesting envars %s excluding %s",
+                            (NULL == pmix_mca_pnet_opa_component.incparms)
+                                ? "NONE"
+                                : pmix_mca_pnet_opa_component.incparms,
+                            (NULL == pmix_mca_pnet_opa_component.excparms)
+                                ? "NONE"
+                                : pmix_mca_pnet_opa_component.excparms);
+        /* harvest envars to pass along */
+        PMIX_CONSTRUCT(&cache, pmix_list_t);
+        if (NULL != pmix_mca_pnet_opa_component.include) {
+            rc = pmix_util_harvest_envars(pmix_mca_pnet_opa_component.include,
+                                          pmix_mca_pnet_opa_component.exclude, &cache);
+            if (PMIX_SUCCESS != rc) {
+                PMIX_LIST_DESTRUCT(&cache);
+                PMIX_DESTRUCT(&mydata);
+                return rc;
+            }
+            /* pack anything that was found */
+            PMIX_LIST_FOREACH (kv, &cache, pmix_kval_t) {
+                PMIX_BFROPS_PACK(rc, pmix_globals.mypeer, &mydata,
+                                 &kv->value->data.envar, 1, PMIX_ENVAR);
+            }
+            PMIX_LIST_DESTRUCT(&cache);
+        }
+    }
+
+    // unload the buffer
+    PMIX_UNLOAD_BUFFER(&mydata, bo.bytes, bo.size);
+    // if nothing was packed, then we have nothing to send
+    if (0 == bo.size) {
+        PMIX_DESTRUCT(&mydata);
+        return PMIX_SUCCESS;
+    }
+
+    /* load all our results into a buffer for xmission to the backend */
+    PMIX_KVAL_NEW(kv, PMIX_PNET_OPA_BLOB);
+    if (NULL == kv || NULL == kv->value) {
+        PMIX_DESTRUCT(&mydata);
+        return PMIX_ERR_NOMEM;
+    }
+    kv->value->type = PMIX_BYTE_OBJECT;
+    /* to help scalability, compress this blob */
+    if (pmix_compress.compress((uint8_t *) bo.bytes, bo.size,
+                               (uint8_t **) &kv->value->data.bo.bytes, &kv->value->data.bo.size)) {
+        kv->value->type = PMIX_COMPRESSED_BYTE_OBJECT;
+    } else {
+        kv->value->data.bo.bytes = bo.bytes;
+        kv->value->data.bo.size = bo.size;
+    }
+    PMIX_DESTRUCT(&mydata);
+    pmix_list_append(ilist, &kv->super);
+    return PMIX_SUCCESS;
+}
+
+/* PMIx_server_setup_local_support calls the "setup_local_network" function.
+ * The Standard requires that this come _after_ the host calls the
+ * PMIx_server_register_nspace function to ensure that any required information
+ * is available to the components. Thus, we have the PMIX_NODE_MAP and
+ * PMIX_PROC_MAP available to us and can use them here.
+ *
+ * When the host calls "setup_local_support", it passes down an array
+ * containing the information the "lead" server (e.g., "mpirun") collected
+ * from PMIx_server_setup_application. In this case, we search for a blob
+ * that our "allocate" function may have included in that info.
+ */
+static pmix_status_t setup_local_network(pmix_nspace_env_cache_t *ns,
+                                         pmix_info_t info[], size_t ninfo)
+{
+    size_t n;
+    pmix_buffer_t bkt;
+    int32_t cnt;
+    pmix_status_t rc = PMIX_SUCCESS;
+    uint8_t *data;
+    size_t size;
+    bool release = false;
+    pmix_envar_list_item_t *ev;
+    pmix_proc_t proc;
+    pmix_kval_t *kv;
+
+    pmix_output_verbose(2, pmix_pnet_base_framework.framework_output,
+                        "pnet:opa:setup_local with %lu info", (unsigned long) ninfo);
+
+    /* prep the unpack buffer */
+    PMIX_CONSTRUCT(&bkt, pmix_buffer_t);
+
+    for (n = 0; n < ninfo; n++) {
+        /* look for my key */
+        if (PMIX_CHECK_KEY(&info[n], PMIX_PNET_OPA_BLOB)) {
+            pmix_output_verbose(2, pmix_pnet_base_framework.framework_output,
+                                "pnet:opa:setup_local found my blob");
+
+            /* if this is a compressed byte object, decompress it */
+            if (PMIX_COMPRESSED_BYTE_OBJECT == info[n].value.type) {
+                pmix_compress.decompress(&data, &size, (uint8_t *) info[n].value.data.bo.bytes,
+                                         info[n].value.data.bo.size);
+                release = true;
+            } else {
+                data = (uint8_t *) info[n].value.data.bo.bytes;
+                size = info[n].value.data.bo.size;
+            }
+            PMIX_LOAD_BUFFER_NON_DESTRUCT(pmix_globals.mypeer, &bkt, data, size);
+
+            /* all we packed was envars, so just cycle thru */
+            ev = PMIX_NEW(pmix_envar_list_item_t);
+            cnt = 1;
+            PMIX_BFROPS_UNPACK(rc, pmix_globals.mypeer, &bkt, &ev->envar, &cnt, PMIX_ENVAR);
+            while (PMIX_SUCCESS == rc) {
+                pmix_list_append(&ns->envars, &ev->super);
+                /* if this is the transport key, save it */
+                if (0 == strncmp(ev->envar.envar, "OMPI_MCA_orte_precondition_transports", PMIX_MAX_KEYLEN)) {
+                    /* add it to the job-level info */
+                    PMIX_LOAD_PROCID(&proc, ns->ns->nspace, PMIX_RANK_WILDCARD);
+                    PMIX_KVAL_NEW(kv, PMIX_CREDENTIAL);
+                    kv->value->type = PMIX_STRING;
+                    kv->value->data.string = strdup(ev->envar.value);
+                    PMIX_GDS_STORE_KV(rc, pmix_globals.mypeer, &proc, PMIX_INTERNAL, kv);
+                    PMIX_RELEASE(kv); // maintain refcount
+                    if (PMIX_SUCCESS != rc) {
+                        PMIX_ERROR_LOG(rc);
+                    }
+                }
+                /* get the next envar */
+                ev = PMIX_NEW(pmix_envar_list_item_t);
+                cnt = 1;
+                PMIX_BFROPS_UNPACK(rc, pmix_globals.mypeer, &bkt, &ev->envar, &cnt, PMIX_ENVAR);
+            }
+            if (PMIX_ERR_UNPACK_READ_PAST_END_OF_BUFFER == rc) {
+                rc = PMIX_SUCCESS;
+            }
+            // we will have created one more envar than we want
+            PMIX_RELEASE(ev);
+
+            /* we are done */
+            break;
+        }
+    }
+
+    if (release) {
+        free(data);
+    }
+
+    return rc;
+}
+
+static pmix_status_t collect_inventory(pmix_info_t directives[], size_t ndirs,
+                                       pmix_list_t *inventory)
+{
+    /* search the topology for OPA NICs */
+    PMIX_HIDE_UNUSED_PARAMS(directives, ndirs, inventory);
+
+
+    return PMIX_SUCCESS;
+}
+
+static pmix_status_t deliver_inventory(pmix_info_t info[], size_t ninfo,
+                                       pmix_info_t directives[], size_t ndirs)
+{
+    /* look for our inventory blob */
+    PMIX_HIDE_UNUSED_PARAMS(info, ninfo, directives, ndirs);
+
+
+    return PMIX_SUCCESS;
+}
diff --git a/src/mca/pnet/opa/pnet_opa.h b/src/mca/pnet/opa/pnet_opa.h
new file mode 100644
index 0000000000..0f06f54b54
--- /dev/null
+++ b/src/mca/pnet/opa/pnet_opa.h
@@ -0,0 +1,41 @@
+/*
+ * Copyright (c) 2015-2020 Intel, Inc.  All rights reserved.
+ *
+ * Copyright (c) 2021-2022 Nanook Consulting.  All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#ifndef PMIX_PNET_OPA_H
+#define PMIX_PNET_OPA_H
+
+#include "src/include/pmix_config.h"
+
+#include "src/mca/pnet/pnet.h"
+
+BEGIN_C_DECLS
+
+typedef struct {
+    pmix_pnet_base_component_t super;
+    char *incparms;
+    char *excparms;
+    char **include;
+    char **exclude;
+} pmix_pnet_opa_component_t;
+
+/* the component must be visible data for the linker to find it */
+PMIX_EXPORT extern pmix_pnet_opa_component_t pmix_mca_pnet_opa_component;
+extern pmix_pnet_module_t pmix_opa_module;
+
+/* define a key for any blob we need to send in a launch msg */
+#define PMIX_PNET_OPA_BLOB "pmix.pnet.opa.blob"
+
+/* define an inventory key */
+#define PMIX_PNET_OPA_INVENTORY_KEY "pmix.pnet.opa.inventory"
+
+END_C_DECLS
+
+#endif
diff --git a/src/mca/pnet/opa/pnet_opa_component.c b/src/mca/pnet/opa/pnet_opa_component.c
new file mode 100644
index 0000000000..8377640913
--- /dev/null
+++ b/src/mca/pnet/opa/pnet_opa_component.c
@@ -0,0 +1,114 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil -*- */
+/*
+ * Copyright (c) 2004-2008 The Trustees of Indiana University and Indiana
+ *                         University Research and Technology
+ *                         Corporation.  All rights reserved.
+ * Copyright (c) 2004-2005 The University of Tennessee and The University
+ *                         of Tennessee Research Foundation.  All rights
+ *                         reserved.
+ * Copyright (c) 2004-2005 High Performance Computing Center Stuttgart,
+ *                         University of Stuttgart.  All rights reserved.
+ * Copyright (c) 2004-2005 The Regents of the University of California.
+ *                         All rights reserved.
+ * Copyright (c) 2015      Los Alamos National Security, LLC. All rights
+ *                         reserved.
+ * Copyright (c) 2016-2020 Intel, Inc.  All rights reserved.
+ * Copyright (c) 2021-2025 Nanook Consulting  All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ *
+ * These symbols are in a file by themselves to provide nice linker
+ * semantics.  Since linkers generally pull in symbols by object
+ * files, keeping these symbols as the only symbols in this file
+ * prevents utility programs such as "ompi_info" from having to import
+ * entire components just to query their version and parameters.
+ */
+
+#include "src/include/pmix_config.h"
+#include "pmix_common.h"
+
+#include "src/hwloc/pmix_hwloc.h"
+#include "src/mca/pnet/base/base.h"
+#include "src/util/pmix_argv.h"
+
+#include "pnet_opa.h"
+
+static pmix_status_t component_open(void);
+static pmix_status_t component_close(void);
+static pmix_status_t component_query(pmix_mca_base_module_t **module, int *priority);
+static pmix_status_t component_register(void);
+
+/*
+ * Instantiate the public struct with all of our public information
+ * and pointers to our public functions in it
+ */
+pmix_pnet_opa_component_t pmix_mca_pnet_opa_component = {
+    .super = {
+        PMIX_PNET_BASE_VERSION_1_0_0,
+
+        /* Component name and version */
+        .pmix_mca_component_name = "opa",
+        PMIX_MCA_BASE_MAKE_VERSION(component,
+                                   PMIX_MAJOR_VERSION,
+                                   PMIX_MINOR_VERSION,
+                                   PMIX_RELEASE_VERSION),
+
+        /* Component open and close functions */
+        .pmix_mca_open_component = component_open,
+        .pmix_mca_close_component = component_close,
+        .pmix_mca_register_component_params = component_register,
+        .pmix_mca_query_component = component_query,
+    },
+    .include = NULL,
+    .exclude = NULL
+};
+
+static pmix_status_t component_register(void)
+{
+    pmix_mca_base_component_t *component = &pmix_mca_pnet_opa_component.super;
+
+    pmix_mca_pnet_opa_component.incparms = "HFI_*,PSM2_*";
+    (void) pmix_mca_base_component_var_register(
+        component, "include_envars",
+        "Comma-delimited list of envars to harvest (\'*\' and \'?\' supported)",
+        PMIX_MCA_BASE_VAR_TYPE_STRING,
+        &pmix_mca_pnet_opa_component.incparms);
+    if (NULL != pmix_mca_pnet_opa_component.incparms) {
+        pmix_mca_pnet_opa_component.include = PMIx_Argv_split(pmix_mca_pnet_opa_component.incparms, ',');
+    }
+
+    pmix_mca_pnet_opa_component.excparms = NULL;
+    (void) pmix_mca_base_component_var_register(
+        component, "exclude_envars",
+        "Comma-delimited list of envars to exclude (\'*\' and \'?\' supported)",
+        PMIX_MCA_BASE_VAR_TYPE_STRING,
+        &pmix_mca_pnet_opa_component.excparms);
+    if (NULL != pmix_mca_pnet_opa_component.excparms) {
+        pmix_mca_pnet_opa_component.exclude = PMIx_Argv_split(pmix_mca_pnet_opa_component.excparms, ',');
+    }
+
+    return PMIX_SUCCESS;
+}
+
+static pmix_status_t component_open(void)
+{
+    pmix_status_t rc;
+
+    rc = pmix_hwloc_check_vendor(&pmix_globals.topology, 0x8086, 0x208);
+    return rc;
+}
+
+static pmix_status_t component_query(pmix_mca_base_module_t **module, int *priority)
+{
+    *priority = 10;
+    *module = (pmix_mca_base_module_t *) &pmix_opa_module;
+    return PMIX_SUCCESS;
+}
+
+static pmix_status_t component_close(void)
+{
+    return PMIX_SUCCESS;
+}
